sql/rls: push leakproof projections through permeable Barriers

spilchen opened the pull request:
Adds a normalization rule to push a Project operator below a Barrier when all projection expressions are leakproof and the Barrier is marked as permeable. This is needed to optimizer RLS queries since those have Barriers added to support the RLS filter. This rule will help bubble the Barrier up the plan tree.

The rule is applied if all projections are leakproof. I wasn't able to split the projection into leakproof and non-leakproof like the PushLeakproofFiltersIntoPermeableBarrier rule does for filter. I didn't know how to split the passthrough columns that go along with the projection.

Informs #146952

Epic: CRDB-48807
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148203)
<!-- Reviewable:end -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148203)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148203#-OSdsi5JDTK6MPrF9YEH:-OSdvYyQ1HnMX2h09b-t:b7i6ouo):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

To split the project, we'd use the original passthrough columns in the inner project, and the outer project would be the union of the original passthrough columns and the columns projected by the inner project. I'm happy to take a shot at this, if you'd like.
</blockquote></details>

But maybe it's worthwhile to ask what benefit we'd expect to see if we did this. Is there a meaningful benefit for real-world queries?
___
*[`pkg/sql/opt/norm/testdata/rules/project` line 2059 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148203#-OSdtwNoCioXyNoF6Ajj:-OSdtwNoCioXyNoF6Ajk:b61w371) ([raw file](https://github.com/cockroachdb/cockroach/blob/601d748006dfab5295664916da3e3edcc5198c4d/pkg/sql/opt/norm/testdata/rules/project#L2059)):*
> ```
> FROM rls
> GROUP BY bucket
> HAVING max(CASE WHEN y = 20 THEN y ELSE 0 END) > 0;
> ```

We typically make test cases as simple as possible while still exercising the intended code path. Is there a reason that we need the `GROUP BY` and `HAVING`? It seems like some tests like this would be good enough:

```
SELECT y+1 FROM rls

SELECT y/0 FROM rls
```


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148203)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @spilchen)

___
*[`-- commits` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148203#-OSdsi5JDTK6MPrF9YEH:-OSekOeWBj3Ex7XsEgQl:b61omuq):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

But maybe it's worthwhile to ask what benefit we'd expect to see if we did this. Is there a meaningful benefit for real-world queries?
</blockquote></details>

I didn't have a real world example that this benefits. I was just following the pattern we did to split up the filters in the previous PR.
___
*[`pkg/sql/opt/norm/testdata/rules/project` line 2059 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148203#-OSdtwNoCioXyNoF6Ajj:-OSekYVo5KCAGC3HkpAF:btd4cfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/601d748006dfab5295664916da3e3edcc5198c4d/pkg/sql/opt/norm/testdata/rules/project#L2059)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

We typically make test cases as simple as possible while still exercising the intended code path. Is there a reason that we need the `GROUP BY` and `HAVING`? It seems like some tests like this would be good enough:

```
SELECT y+1 FROM rls

SELECT y/0 FROM rls
```
</blockquote></details>

I may have relied on ChatGPT a bit too much here to generate the tests. I think that's where the GROUP BY came from. I used your simpler suggestion. Although, y+1 is not considered leakproof I believe. So, I kept the CASE statement, just didn't end up using the GROUP BY.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just a few minor suggestions.

> Reviewed 4 of 5 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148203#-:-OSznNKUAhPYJ7n_ERXa:bcdwac6)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/norm/testdata/rules/project` line 2059 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148203#-OSdtwNoCioXyNoF6Ajj:-OSzmVpkDH53HtQhNcSk:byimgvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/601d748006dfab5295664916da3e3edcc5198c4d/pkg/sql/opt/norm/testdata/rules/project#L2059)):*
> Although, y+1 is not considered leakproof I believe.

Ah right—it can error on overflow.
___
*[`pkg/sql/opt/norm/testdata/rules/project` line 2056 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148203#-OSzmf6X7mtwa2HlJSFl:-OSzmf6X7mtwa2HlJSFm:bixka76) ([raw file](https://github.com/cockroachdb/cockroach/blob/59609240f6af699ae85696a0d46f9b528fd85daf/pkg/sql/opt/norm/testdata/rules/project#L2056)):*
> ```
> # All expressions in projection are leakproof
> norm expect=PushLeakproofProjectionsIntoPermeableBarrier
> select case when y = 20 then 10 else 0 end FROM rls
> ```

nit: capitalize SQL keywords here and below
___
*[`pkg/sql/opt/norm/testdata/rules/project` line 2056 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148203#-OSzn0Q05xQrwSPWjOE8:-OSzn0Q05xQrwSPWjOE9:b-u41fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/59609240f6af699ae85696a0d46f9b528fd85daf/pkg/sql/opt/norm/testdata/rules/project#L2056)):*
> ```
> # All expressions in projection are leakproof
> norm expect=PushLeakproofProjectionsIntoPermeableBarrier
> select case when y = 20 then 10 else 0 end FROM rls
> ```

Can you add a test case with some passthrough columns, e.g., `select x, y, case when y = 20 then 10 else 0 end FROM rls`?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/rls: push leakproof projections through permeable Barriers

Adds a normalization rule to push a Project operator below a Barrier
when all projection expressions are leakproof and the Barrier is
marked as permeable. This is needed to optimize RLS queries since
those have Barriers added to support the RLS filter. This rule will
help bubble the Barrier up the plan tree.

The rule is applied if all projections are leakproof. I didn't split the
projection into leakproof and non-leakproof like the
PushLeakproofFiltersIntoPermeableBarrier rule does for filter.

Informs #146952

Epic: CRDB-48807
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148203)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/norm/testdata/rules/project` line 2056 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148203#-OSzn0Q05xQrwSPWjOE8:-OT1lrJuEYF9XlJFO7QU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59609240f6af699ae85696a0d46f9b528fd85daf/pkg/sql/opt/norm/testdata/rules/project#L2056)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you add a test case with some passthrough columns, e.g., `select x, y, case when y = 20 then 10 else 0 end FROM rls`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
